pkg/build/ignore/docker/dockign.go

```diff
@@ -0,0 +1,135 @@
+package docker
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/openshift/source-to-image/pkg/api"
+	"fmt"
+)
+
+type DockIgn struct {
+}
+
+func (b *DockIgn) Ignore(config *api.Config) error {
+	// so, to duplicate the .dockerignore capabilities (https://docs.docker.com/reference/builder/#dockerignore-file)
+	// we have a flow that follows:
+	/*
+
+		0) First note, .dockerignore rules are NOT recursive (unlike .gitignore) .. you have to list subdir explicitly
+		1) Read in the exclusion patterns
+		2) Skip over comments (noted by #)
+		3) note overrides (via exclamation sign i.e. !) and reinstate files (don't remove) as needed
+		4) leverage Glob matching to build list, as .dockerignore is documented as following filepath.Match / filepath.Glob
+		5) del files
+
+                 1 to 4 is in getListOfFilesToIgnore
+
+	*/
+
+	filesToDel, lerr := getListOfFilesToIgnore(config)
+	if lerr != nil {
+		return lerr
+	}
+
+	if filesToDel == nil {
+		return nil
+	}
+
+	// delete compiled list of files 
+	for _, fileToDel := range filesToDel {
+		glog.V(5).Infof("attempting to remove file %s \n", fileToDel)
+		rerr := os.RemoveAll(fileToDel)
+		if rerr != nil {
+			glog.Errorf("error removing file %s because of %v \n", fileToDel, rerr)
+			return rerr
+		}
+	}
+
+	return nil
+}
+
+func getListOfFilesToIgnore(config *api.Config) (map[string]string, error) {
+	path := filepath.Join(config.WorkingSourceDir, api.DefaultIgnoreFile)
+	file, err := os.Open(path)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			glog.Errorf("Ignore processing, problem opening %s because of %v\n", path, err)
+			return nil, err
+		} else {
```

Comment:
Your tears give me strength

On Jul 28, 2015, at 10:22 AM, Ben Parees <notifications@github.com> wrote:

In pkg/build/ignore/docker/dockign.go
<https://github.com/openshift/source-to-image/pull/248#discussion_r35652714>
:

> +			glog.Errorf("error removing file %s because of %v \n", fileToDel, rerr)
> +			return rerr
> +		}
> +	}
> +
> +	return nil
> +}
> +
> +func getListOfFilesToIgnore(config *api.Config) (map[string]string, error) {
> +	path := filepath.Join(config.WorkingSourceDir, api.DefaultIgnoreFile)
> +	file, err := os.Open(path)
> +	if err != nil {
> +		if !os.IsNotExist(err) {
> +			glog.Errorf("Ignore processing, problem opening %s because of %v\n", path, err)
> +			return nil, err
> +		} else {

no need for this else block since you're returning, just put the return
outside the if.

@smarterclayton <https://github.com/smarterclayton> i hope you're proud of
yourself for what you've turned me into.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/source-to-image/pull/248/files#r35652714>.